doc: Add a repo-level README (#169)
googleapis/mcp-toolbox-sdk-python: README.md

as standard Python functions or objects within popular orchestration frameworks
or your custom code.

This simplifies the process of incorporating external functionalities (like
Databases or APIs) managed by Toolbox into your GenAI applications.

<!-- TOC -->

- [Overview](#overview)
- [Available Packages](#available-packages)
- [Quickstart](#quickstart)
- [Core Concepts](#core-concepts)
    - [Connecting to Toolbox](#connecting-to-toolbox)
    - [Loading Tools](#loading-tools)
    - [Invoking Tools](#invoking-tools)
    - [Synchronous vs. Asynchronous Usage](#synchronous-vs-asynchronous-usage)
    - [Authenticating Tools](#authenticating-tools)
        - [When is Authentication Needed?](#when-is-authentication-needed)
        - [Supported Authentication Mechanisms](#supported-authentication-mechanisms)
        - [SDK Configuration](#sdk-configuration)
    - [Binding Parameter Values](#binding-parameter-values)
        - [Why Bind Parameters?](#why-bind-parameters)
        - [SDK Configuration](#sdk-configuration)
- [Usage Comparison](#usage-comparison)
- [Contributing](#contributing)
- [License](#license)
- [Support](#support)

<!-- /TOC -->

> ## Overview

Reviewer: nit: it shows 2 TOCs

Author: Oops...the plugin I used seems to act a bit funny though. Thanks for pointing that out.

The Toolbox service provides a centralized way to manage and expose tools for
use by LLMs. <!-- TOC -->

- [MCP Toolbox SDKs for Python](#mcp-toolbox-sdks-for-python)
    - [Overview](#overview)
    - [Available Packages](#available-packages)
    - [Quickstart](#quickstart)
    - [Core Concepts](#core-concepts)
        - [Connecting to Toolbox](#connecting-to-toolbox)
        - [Loading Tools](#loading-tools)
        - [Invoking Tools](#invoking-tools)
        - [Synchronous vs. Asynchronous Usage](#synchronous-vs-asynchronous-usage)
        - [Authenticating Tools](#authenticating-tools)
            - [When is Authentication Needed?](#when-is-authentication-needed)
            - [Supported Authentication Mechanisms](#supported-authentication-mechanisms)
            - [SDK Configuration](#sdk-configuration)
        - [Binding Parameter Values](#binding-parameter-values)
            - [Why Bind Parameters?](#why-bind-parameters)
            - [SDK Configuration](#sdk-configuration)
    - [Usage Comparison](#usage-comparison)
    - [Contributing](#contributing)
    - [License](#license)
    - [Support](#support)

> <!-- /TOC -->These SDKs act as clients for that service, abstracting away the

Reviewer: nit: please correcting formatting. see [this](https://screenshot.googleplex.com/AhttT4S3fs3RMdq)

### Connecting to Toolbox

Initialize a client, pointing it to the URL where your Toolbox service is
running.

```py
from toolbox_core import ToolboxClient

# replace with your Toolbox service's URL
client = ToolboxClient("http://127.0.0.1:5000")
```

### Loading Tools

Fetch tool definitions from the Toolbox service. You can load individual tools
by name or load all tools within a specific toolset (or all available toolsets).

```py
# Load a single tool
tool = await client.load_tool("my-tool")

# Load all tools in a specific toolset
tools = await client.load_toolset("my-toolset")

# Load all tools from all toolsets
all_tools = await client.load_toolset()
```

> [!NOTE]
> **Upcoming Feature (`v0.2.0`+):** Support for calling `toolbox.load_toolset()`
> without arguments will be introduced in `v0.2.0`.

### Invoking Tools

Loaded tools behave like callable Python objects or functions.

  * **`toolbox-core`**: Async tools are `awaitable`, sync tools are called
    directly.
  * **`toolbox-langchain`**: Tools conform to LangChain's `BaseTool` interface
    and are typically invoked via `.invoke()` or `.ainvoke()`, often managed by
    a LangGraph agent.

```py

> # toolbox-core

Reviewer: Let's move the code under the correct bullet points?

Author: Good idea. Thanks!

| **`toolbox-langchain`** | **Unified Class:** `ToolboxClient`<br/><br/>* **Async Methods:** `await client.aload_*`<br/><br/>* **Sync Methods:** `client.load_*` | **Unified Class:** `ToolboxTool`<br/><br/>* &nbsp;**Async Method:** `await tool.ainvoke()`<br/><br/>* &nbsp;**Sync Method:** `tool.invoke()`                                                   | [`toolbox-langchain` README](https://github.com/googleapis/mcp-toolbox-sdk-python/blob/main/packages/toolbox-langchain/README.md#asynchronous-usage) |

The `ToolboxSyncClient` handles communication with the Toolbox service *synchronously* and produces `ToolboxSyncTool` instances when you load tools. You do not use the `await` keyword when interacting with these synchronous versions.

```py
from toolbox_core import ToolboxSyncClient

# replace with your Toolbox service's URL
sync_client = ToolboxSyncClient("http://127.0.0.1:5000")

# Load a single tool
tool = sync_client.load_tool("my-tool")

# Load all tools in a specific toolset
tools = sync_client.load_toolset("my-toolset")

# Load all tools from all toolsets

> all_tools = sync_client.load_toolset()

Reviewer: This does not work yet. Should we add a note?

  * **Pre-filling known data:** Pre-fill common or known values.

#### SDK Configuration

> [!IMPORTANT]
> The parameter names used for binding must exactly match the parameter names
> defined in the tool's configuration within the Toolbox service.

> [!NOTE]
> You do not need to modify the tool's configuration in the Toolbox service to
> bind parameter values using the SDK.

Similar to authentication, you can bind parameters after loading a tool or
during the loading process.

```py
from toolbox_core import ToolboxClient
# from toolbox_langchain import ToolboxClient

toolbox = ToolboxClient("http://127.0.0.1:5000")
tool = await toolbox.load_tool("my-tool")
# tool = await toolbox.aload_tool("my-tool")

bound_tool = tool.bind_parameters({"param": "value"})
# bound_tool = tool.bind_params({"param": "value"})

# OR

bound_tool = await toolbox.load_tool("my-tool", bound_params={"param": "value"})
# bound_tool = await toolbox.aload_tool("my-tool", bound_params={"param": "value"})
```

## Usage Comparison

While the core concepts are similar, the way you integrate and use the tools
varies depending on the chosen SDK package and framework.

> | Feature                 | `toolbox-core`                                                                                                   | `toolbox-langchain`                                                                        |

Reviewer: I think it will be hard to update this table in the future with any other orchestration sdks. I would prefer if each row corresponds to a package and a column corresponds to a feature so that it's easier to extend. What do you think?

Author: That's a good idea. Thanks!

| **Detailed Examples** | [Core README Section](https://github.com/googleapis/mcp-toolbox-sdk-python/tree/main/packages/toolbox-core#use-with-langgraph)     | [LangChain README Section](https://github.com/googleapis/mcp-toolbox-sdk-python/blob/main/packages/toolbox-langchain#use-with-langgraph) |

## Contributing

Contributions are welcome! Please refer to the
[`CONTRIBUTING.md`](https://github.com/googleapis/mcp-toolbox-sdk-python/blob/main/CONTRIBUTING.md)
to get started.

## License

This project is licensed under the Apache License 2.0. See the
[LICENSE](https://github.com/googleapis/genai-toolbox/blob/main/LICENSE) file
for details.

## Support

If you encounter issues or have questions, please check the existing [GitHub
Issues](https://github.com/googleapis/genai-toolbox/issues) for the main Toolbox
project. If your issue is specific to one of the SDKs, please look for existing
issues [here](https://github.com/googleapis/mcp-toolbox-sdk-python/issues) or

> open a new issue in that repository.

Reviewer: ```suggestion
open a new issue in this repository.
```

![MCP Toolbox
Logo](https://raw.githubusercontent.com/googleapis/genai-toolbox/main/logo.png)
# MCP Toolbox SDKs for Python

[![License: Apache
2.0](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)

> [![PyPI - Python

Reviewer: We're supporting python 3.13 as well. This shows 3.9 to 3.12 only.

Author: Added python 3.13 support in the metadata.

> bind parameter values using the SDK.

Similar to authentication, you can bind parameters after loading a tool or
during the loading process.

```py
from toolbox_core import ToolboxClient

toolbox = ToolboxClient("http://127.0.0.1:5000")
tool = await toolbox.load_tool("my-tool")

bound_tool = tool.bind_parameters({"param": "value"})

# OR

bound_tool = await toolbox.load_tool("my-tool", bound_params={"param": "value"})
```

## Framework-Specific Usage

While the core concepts are similar, the way you integrate and use the tools
varies depending on the chosen SDK package and framework.

### Using `toolbox-core`

  * Ideal for framework-agnostic applications or custom orchestration logic.
  * Use `ToolboxClient` (async) or `ToolboxSyncClient` (sync).
  * Loaded tools (`ToolboxTool`/`ToolboxSyncTool`) are directly
    callable/awaitable.
  * For integration with frameworks like LangGraph that expect specific tool
    formats (e.g., with parsed docstrings for LLM use), you might need to wrap
    the loaded tools (e.g., using LangChain's
    `StructuredTool.from_function(tool, parse_docstring=True)` as shown in the

>     `toolbox-core` README).

Reviewer: ```suggestion
    `toolbox-core` README). Therefore, it is preferred to use the Orchestration framework specific SDKs when available. See all [available packages](https://github.com/googleapis/mcp-toolbox-sdk-python/blob/main/README.md#available-packages)
```

    for detailed examples.

### Using `toolbox-langchain`

  * Designed for seamless use within the LangChain/LangGraph ecosystem.
  * Use `ToolboxClient` (provides both sync `load_*` and async `aload_*`
    methods).
  * Loaded tools are LangChain `BaseTool` compatible objects.
  * Directly usable with LangGraph agents (`model.bind_tools(tools)` and
    `ToolNode(tools)`).
  * See the [toolbox-langchain
    README](https://github.com/googleapis/mcp-toolbox-sdk-python/blob/main/packages/toolbox-langchain#use-with-langgraph)
    for specific LangGraph integration examples.

## Contributing

Contributions are welcome! Please refer to the
[`CONTRIBUTING.md`](https://github.com/googleapis/mcp-toolbox-sdk-python/blob/main/CONTRIBUTING.md)
to get started.

## License

This project is licensed under the Apache License 2.0. See the
[LICENSE](https://github.com/googleapis/genai-toolbox/blob/main/LICENSE) file
for details.

## Support

If you encounter issues or have questions, please check the existing [GitHub
Issues](https://github.com/googleapis/genai-toolbox/issues) for the main Toolbox

> project. If your issue is specific to one of the SDKs and not found, feel free

Reviewer: ```suggestion
project. If your issue is specific to one of the SDKs, please look for existing issues [here](https://github.com/googleapis/mcp-toolbox-sdk-python/issues) or 
```

> to open a new issue in that repository.

Reviewer: ```suggestion
open a new issue in that repository.
```
